pkg/kgo/compression.go

```diff
@@ -202,7 +194,9 @@ func (c *compressor) compress(dst *sliceWriter, src []byte, produceRequestVersio
 		}
 
 	case codecSnappy:
-		dst.inner = s2.EncodeSnappy(dst.inner[:cap(dst.inner)], src)
+		out := s2.EncodeSnappy(dst.Bytes(), src)
+		dst.Reset()
```

Comment:
This chunk of code _might_ actually be a reason to keep sliceWriters as well.

* EncodeSnappy (and below in zstd, EncodeAll) both take an input slice, and _if there is capacity_, uses it, _if not_, returns a new larger slice
* Now, the code doubly writes and potentially doubly allocates. If Encode returns a larger slices, dst.Write(out) actually also reallocates the internal slice. Even if it does not reallocate, the Write is an unnecessary copy.

One way to hack around this in the current logic is to change the function signature and to check:

```go
 func (c *compressor) compress(dst *bytes.Buffer, 

// ... other code

if len(out) == 0 {
  // skip
}
dst.Reset()
dstBytes := dst.Bytes()
dstBytes = dstBytes[:cap(dstBytes)]
if len(dstBytes) == 0 || &dstBytes[0] != &out[0] {
  // replace the bytes.Buffer itself, discard old bytes.Buffer
  *dstp = bytes.NewBuffer(dstBytes)
  dst = *dstp
}
```

^^Essentially, keep the old bytes.Buffer if the slice didn't change, otherwise replace the bytes.Buffer with a new one if the slice did change. It's really convoluted though, and this may be a reason to keep/use sliceWriters. wdyt?

Reply:
Maybe we can utilize the methods in zstd & snappy that provide the length of expected decompressed bytes and grow the byte buffer accordingly beforehand? These methods are exported by the libs and these algos probably want us to leverage this info. This way we can be sure the code won't doubly allocate and neither copy again

![image](https://github.com/twmb/franz-go/assets/14031096/a52c8c44-2370-4b11-81b8-7be260edaed7)

Reply:
Great idea w/ MaxEncodedLen and MaxEncodedSize, I didn't know those methods existed on these compressors